testing/unit_testing/count_cohorts_test/test_CountCohorts.pf

```diff
@@ -0,0 +1,98 @@
+module test_CountCohorts
+  !
+  ! DESCRIPTION:
+  !		Tests the patch's CountCohorts method
+  !
+  use FatesConstantsMod, only : r8 => fates_r8
+  use FatesCohortMod,    only : fates_cohort_type
+  use FatesPatchMod,     only : fates_patch_type
+  use funit
+
+  implicit none
+  
+  @TestCase
+  type, extends(TestCase) :: TestCountCohorts
+    type(fates_patch_type), pointer :: patch
+    contains
+      procedure :: setUp
+      procedure :: tearDown
+  end type TestCountCohorts
+
+  real(r8), parameter :: tol = 1.e-13_r8
+
+  contains 
+  
+    subroutine setUp(this)
+      class(TestCountCohorts), intent(inout) :: this           ! test object
+      type(fates_cohort_type), pointer       :: new_node, head ! cohort objects
+      
+      ! heights for cohors
+      real(r8), parameter :: heights(8) = (/2.0_r8, 5.0_r8, 10.0_r8, 12.0_r8, 12.5_r8,   &
+         15.0_r8, 20.0_r8, 25.0_r8/)
+
+      ! hard-code a linked list
+      allocate(this%patch)
+      
+      allocate(head)
+      this%patch%shortest => head
+      head%height = heights(1)
+      
+      allocate(new_node)
+      new_node%height = heights(2)
+      head%taller => new_node
+      new_node%shorter => head
+      
+      allocate(new_node)
+      new_node%height = heights(3)
+      head%taller%taller => new_node
+      new_node%shorter => head%taller
+      
+      allocate(new_node)
+      new_node%height = heights(4)
+      head%taller%taller%taller => new_node
+      new_node%shorter => head%taller%taller
+      
+      allocate(new_node)
+      new_node%height = heights(5)
+      head%taller%taller%taller%taller => new_node
+      new_node%shorter => head%taller%taller%taller
+      
+      allocate(new_node)
+      new_node%height = heights(6)
+      head%taller%taller%taller%taller%taller => new_node
+      new_node%shorter => head%taller%taller%taller%taller
+      
+      allocate(new_node)
+      new_node%height = heights(7)
+      head%taller%taller%taller%taller%taller%taller => new_node
+      new_node%shorter => head%taller%taller%taller%taller%taller
+      
+      allocate(new_node)
+      new_node%height = heights(8)
+      head%taller%taller%taller%taller%taller%taller%taller => new_node
+      new_node%shorter => head%taller%taller%taller%taller%taller%taller
+      this%patch%tallest => new_node
+      
+    end subroutine setUp
+    
+    subroutine tearDown(this)
+      class(TestCountCohorts), intent(inout) :: this        ! test object
+      type(fates_cohort_type), pointer       :: cohort      ! cohort object
+      type(fates_cohort_type), pointer       :: next_cohort ! next cohort object
+      
+      ! deallocate cohorts
```

Comment:
I have never used the finalization feature of Fortran (keyword `final`), but this seems like a possibly-appropriate use for it – i.e., introducing a finalization routine for `fates_patch_type` that deallocates the cohort linked list, so that when tests (or maybe production code, too) deallocates a patch, this underlying cohort deallocation happens automatically (if I understand finalization right).

But that's just an idea - not something you necessarily have to do.

Reply:
oh interesting. I had thought about making some `final` subroutines for the patch when I did that refactor. I think it is a good idea. I have also never used that feature. There are some other things we deallocate on the production-code patch (arrays, other classes) as well. Perhaps this should be another PR?

Reply:
Yes, deferring this to later work makes sense.